src/Test.UnitTests.Sarif/Visitors/RunMergingVisitorTests.cs

```diff
@@ -359,5 +360,90 @@ private void VerifyRuleMatches(Run mergedRun, Run sourceRun, int firstResultInde
+                    new Result { RuleId = "Rule/002/SubRule2", RuleIndex = 1 },
+                    new Result { RuleId = "Rule004", RuleIndex = 3 }, // not a sub rule
+                }
+            };
+
+            var currentRun = new Run
+            {
+                Tool = new Tool
+                {
+                    Driver = new ToolComponent
+                    {
+                        Name = "Test Tool",
+                        Rules = new List<ReportingDescriptor>()
+                        {
+                            new ReportingDescriptor() { Id = "Rule/004" },
+                            new ReportingDescriptor() { Id = "Rule/002" },
+                            new ReportingDescriptor() { Id = "Rule/003" },
+                            new ReportingDescriptor() { Id = "Rule/005" },
+                        }
+                    },
+                },
+                Results = new List<Result>
+                {
+                    new Result { RuleId = "Rule/002/SubRule3", RuleIndex = 1 },
+                    new Result { RuleId = "Rule/002/SubRule4", RuleIndex = 1 },
+                    new Result { RuleId = "Rule/003/SubRule1", RuleIndex = 2 },
+                    new Result { RuleId = "Rule/003/SubRule2", RuleIndex = 2 },
+                    new Result { RuleIndex = 3 }, // RuleId is not set ("Rule/005")
+                    new Result { RuleId = "Rule/DoesNotExist" }, // rule definition does not exist, should create a empty rule for it.
+                }
+            };
+
+            Run mergedRun = currentRun.DeepClone();
+            Run baselineRunCopy = previousRun.DeepClone();
+
+            mergedRun.MergeResultsFrom(baselineRunCopy);
+
+            // We should get Rule/001, Rule/002, Rule/003, Rule/004 and Rule/005.
+            // Rule/006 wasn't referenced. With 1 extra rule "Rule/DoesNotExist".
+            mergedRun.Tool.Driver.Rules.Count.Should().Be(6);
+            mergedRun.Tool.Driver.Rules.Any(r => r.Id == "Rule/001").Should().BeTrue();
+            mergedRun.Tool.Driver.Rules.Any(r => r.Id == "Rule/002").Should().BeTrue();
+            mergedRun.Tool.Driver.Rules.Any(r => r.Id == "Rule/003").Should().BeTrue();
+            mergedRun.Tool.Driver.Rules.Any(r => r.Id == "Rule004").Should().BeTrue();
+            mergedRun.Tool.Driver.Rules.Any(r => r.Id == "Rule/005").Should().BeTrue();
+            mergedRun.Tool.Driver.Rules.Any(r => r.Id == "Rule/006").Should().BeFalse();
+            mergedRun.Tool.Driver.Rules.Any(r => r.Id == "Rule/DoesNotExist").Should().BeTrue();
+
+            mergedRun.Results.Count.Should().Be(11);
+            foreach (Result result in mergedRun.Results)
+            {
+                string subRuleId = result.RuleId;
+                string parentRuleId = result.GetRule(mergedRun).Id;
+                if (subRuleId != parentRuleId)
+                {
+                    subRuleId.StartsWith(parentRuleId).Should().BeTrue();
+                }
+            }
```

Comment:
We can consolidate all of this:
```
int rule_001Count = 0;
int rule_002Count = 0;
int rule_003Count = 0;
int rule004Count = 0;
int rule_005Count = 0;
int rule_006Count = 0;
int rule_DoesNotExistCount = 0;

mergedRun.Results.Count.Should().Be(11); //Even better if you have a way of calculating this expected value from the inputs above.

foreach (Result result in mergedRune.Results)
{
    ...// keep everything the same and add something like this:
   if(result.Id == "Rule/001")
   {
       rule_001Count+=1;
   }
}

rule_001Count.Should().Be(expectedRule001Count);
//etc
```

We're trying to avoid iterating over the same lists multiple times. Let me know if this makes sense.

Reply:
agree, updated & removed duplicated iterations